pytorch_lightning/utilities/auto_restart.py

```diff
@@ -207,29 +207,189 @@ def __iter__(self) -> Iterator:
+        # move back to 0
+        while next(iter_dataloader._worker_queue_idx_cycle) != 0:
+            pass
+        # increment previous worker
+        if isinstance(state_dict["previous_worker"], int):
+            for _ in range(state_dict["previous_worker"] - 1):
+                next(iter_dataloader._worker_queue_idx_cycle)
+        iter_dataloader._reset = iter_dataloader._original_reset
+        iter_dataloader._reset(dataloader, first_iter=True)
+    return iter_dataloader
+
+
+def dataloader_to_state_dict(
+    dataloader: DataLoader,
+    iterator: Iterator,
+    num_batches_processed: int = None,
+) -> List[Dict[str, Any]]:
+    """
+    Convert a dataloader to its associated state dict
+    """
+    out = {}
+    if iterator is not None:
+        out.update(find_current_worker(iterator))
+
+    if not isinstance(dataloader.dataset, CaptureIterableDataset):
+        fast_forward_sampler = find_fast_forward_samplers(dataloader)
+        if fast_forward_sampler is not None:
+            out.update(fast_forward_sampler.state_dict(num_batches_processed=num_batches_processed))
+    return out
+
+
+def dataloader_load_state_dict(dataloader: DataLoader, state_dict: List[Dict[str, Any]]) -> DataLoader:
+    """
+    Reload ``DataLoader`` fast-forward sampler state dict.
+    """
+    fast_forward_sampler = find_fast_forward_samplers(dataloader)
+
+    if isinstance(fast_forward_sampler, Sampler):
+        state_dict = {k: v for k, v in state_dict.items() if k not in ("num_workers", "previous_worker")}
+        fast_forward_sampler.load_state_dict(state_dict)
+
+    return dataloader
+
+
+def find_current_worker(iterator: Iterator) -> Dict[str, Optional[int]]:
+    """
+    Find the current DataLoader Iterator worker if multiple workers were used..
+    """
+    num_workers = getattr(iterator, "_num_workers", 0)
+    if isinstance(iterator, _MultiProcessingDataLoaderIter):
+        next_worker = (next(iterator._worker_queue_idx_cycle)) % num_workers
+        previous_worker = (next_worker - 1) % num_workers
+        while next(iterator._worker_queue_idx_cycle) != previous_worker:
+            pass
+    else:
+        previous_worker = None
+
+    return {"num_workers": num_workers, "previous_worker": previous_worker}
+
+
+def sampler_metadata_collate(samples: List[Any], dataset, default_collate):
+    """
+    A collate function that adds the state dict of all samplers used in the worker processes.
+
+    The structure will be:
+
+    .. code-block:: python
+
+        {
+            "data": data returned by Dataset
+            "__pl_samplers": {
+                "sampler_name0": state_dict
+                "sampler_name1": state_dict
             }
-        return state_dict
+        }
+    """
+    batch = default_collate(samples)
+    if not isinstance(dataset, CaptureIterableDataset):
+        return batch
+    return {"data": batch, AutoRestartBatchKeys.PL_SAMPLERS: {k: v.state_dict() for k, v in dataset.samplers.items()}}
```

Comment:
Should this call `dataset.state_dict()` instead?

Reply:
Yes, can do :)

Reply:
Done !